app/services/optIn/OptInService.scala

```diff
@@ -42,6 +48,115 @@ class OptInService @Inject()(itsaStatusUpdateConnector: ITSAStatusUpdateConnecto
+                              hc: HeaderCarrier,
+                              ec: ExecutionContext): Future[Option[OptInProposition]] = {
+
+    val savedOptInProposition = for {
+
+      optInSessionData <- OptionT(fetchSavedOptInSessionData())
+      contextData <- OptionT(Future.successful(optInSessionData.optInContextData))
+
+      currentYearAsTaxYear <- OptionT(Future.successful(contextData.currentYearAsTaxYear()))
+      nextTaxYearAsTaxYear <- OptionT(Future.successful(contextData.nextTaxYearAsTaxYear()))
+
+      currentYearITSAStatus = stringToStatus(contextData.currentYearITSAStatus)
+      nextYearITSAStatus = stringToStatus(contextData.nextYearITSAStatus)
+
+      optInInitialState = OptInInitialState(currentYearITSAStatus, nextYearITSAStatus)
+      proposition = createOptInProposition(currentYearAsTaxYear, nextTaxYearAsTaxYear, optInInitialState)
+
+    } yield proposition
+
+    savedOptInProposition.value
+  }
+
+
+  def fetchOptInProposition()(implicit user: MtdItUser[_],
+                               hc: HeaderCarrier,
+                               ec: ExecutionContext): Future[OptInProposition] = {
+
+    fetchSavedOptInProposition().flatMap { savedProposition =>
+      savedProposition.map(Future.successful).getOrElse {
+        val currentYear = dateService.getCurrentTaxYear
+        val nextYear = currentYear.nextYear
+        fetchOptInInitialState(currentYear, nextYear)
+          .map(initialState => createOptInProposition(currentYear, nextYear, initialState))
+      }
+    }
+  }
+
+  private def fetchOptInInitialState( currentYear: TaxYear,
+                                      nextYear: TaxYear)
+                                     (implicit user: MtdItUser[_],
+                                      hc: HeaderCarrier,
+                                      ec: ExecutionContext): Future[OptInInitialState] = {
+
+    val statusMapFuture: Future[Map[TaxYear, ITSAStatus]] = getITSAStatusesFrom(currentYear)
+
+    for {
+      statusMap <- statusMapFuture
+    } yield OptInInitialState(statusMap(currentYear), statusMap(nextYear))
+  }
+
+  private def getITSAStatusesFrom(currentYear: TaxYear)(implicit user: MtdItUser[_],
+                                                         hc: HeaderCarrier,
+                                                         ec: ExecutionContext): Future[Map[TaxYear, ITSAStatus]] =
+    itsaStatusService.getStatusTillAvailableFutureYears(currentYear).map(_.view.mapValues(_.status).toMap.withDefaultValue(ITSAStatus.NoStatus))
+
+  private def createOptInProposition( currentYear: TaxYear,
+                                      nextYear: TaxYear,
+                                      initialState: OptInInitialState
+                                     ): OptInProposition = {
+
+    val currentOptInTaxYear = CurrentOptInTaxYear(
+      status = initialState.currentYearItsaStatus, //todo reinstate after ATs setup ITSAStatus.Annual, //
```

Comment:
I'm not sure I understand these TODO comments. Are they still relevant?

Reply:
comment updated

Reply:
Cool. In that case we just need to get to the bottom of that before merging.

Reply:
done